sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -309,7 +328,8 @@ case class FileSourceScanExec(
           files.map(_.getPath.getName).groupBy(file => BucketingUtils.getBucketId(file))
         val singleFilePartitions = bucketToFilesGrouping.forall(p => p._2.length <= 1)
 
-        if (singleFilePartitions) {
+        // TODO Sort order is currently ignored if buckets are coalesced.
```

Comment:
I can work on preserving the sort order as a separate PR if this PR makes sense.

Reply:
Is this related to https://issues.apache.org/jira/browse/SPARK-24528 ?

Reply:
Yes. This will eliminate the sort when there are multiple files per bucket (and if each file is sorted).

And it seems that the work will be orthogonal to this PR anyway, so I will give it a shot and create a PR if I see some improvements in benchmark.

Reply:
Could you add the jira number in TODO?

Reply:
Added.